### PR TITLE
[IMP] payment: restrict access to `payment.transaction` records

### PR DESCRIPTION
**Pull request by hote-odoo**
[IMP] payment: restrict access to `payment.transaction` records

This commit esentially improve on two points of access rules for `payment.transaction`:
 - before this commit the access rules allows to read your transactions
   and those without a partner -> this no longer will be the case;
 - `payment.transaction` records will only be accessible by default to
    accounting groups, this will make more evident when other users need
    access to transactions by making the use of `sudo` necessary.
    
See odoo/enterprise#40939
    

Task - 3102824

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by chevalierv — APPROVED** [review]
LGTM :+1:

**Comment by pimodoo** [comment]
@hote-odoo @chevalierv Hello,

You are removing 2 groups in the XML, but these records are in "no update", so they won't be deleted during the upgrade. You should do a migration script to remove them.

I don't exactly know how it'll be handled for the access rules, but it could be great to check that they are also correctly deleted. (I think they are)

File: addons/account_payment/security/ir_rules.xml [hunk 1]
```diff
@@ -1,16 +1,6 @@
...
 
-    <!-- Transactions -->
-
-    <record id="payment_transaction_billing_rule" model="ir.rule">
-        <field name="name">Access every transaction</field>
-        <field name="model_id" ref="payment.model_payment_transaction"/>
-        <!-- Reset the domain defined by payment.transaction_user_rule -->
-        <field name="domain_force">[(1, '=', 1)]</field>
-        <field name="groups" eval="[(4, ref('account.group_account_invoice'))]"/>
-    </record>
```
Comment (AntoineVDV) [inline addons/account_payment/security/ir_rules.xml:12]: As Pierre pointed out, you must write an upgrade script to force remove `ir.rule` records are they are flagged as `noupdate`.

It's not needed for `ir.model.access`, though.

File: addons/sale/models/sale_order.py [hunk 2]
```diff
@@ -1349,7 +1349,9 @@ def payment_action_void(self):
 
     def get_portal_last_transaction(self):
         self.ensure_one()
-        return self.transaction_ids._get_last()
+        # In sudo mode to allow the partner of the SO to see transaction
+        # information related to the SO.
+        return self.transaction_ids.sudo()._get_last()
```
Comment (AntoineVDV) [inline addons/sale/models/sale_order.py:1354]: I'm not a fan of `sudo` being used from the models. I'd rather have the `sudo` in the caller and not in the callee so that another model method doesn't inadvertently receive a sudoed recordset, and we can better delimit the scope of the sudo.

In this case, is it possible to move the sudo call in the calling template(s)?

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

File: addons/payment/models/payment_transaction.py [hunk 3]
```diff
@@ -1129,4 +1129,6 @@ def _get_last(self):
         :return: The last transaction of the recordset, sorted by id.
         :rtype: recordset of `payment.transaction`
         """
-        return self.filtered(lambda t: t.state != 'draft').sorted()[:1]
+        if self:
+            return self.filtered(lambda t: t.state != 'draft').sorted()[:1]
+        return self.env['payment.transaction']
```
Comment (hote-odoo) [inline addons/payment/models/payment_transaction.py:1134]: In a call to _get_last the recordset might be empty (partner is then set to false) and access rules will fail.

File: addons/payment/models/payment_transaction.py [hunk 4]
```diff
@@ -1129,4 +1129,4 @@ def _get_last(self):
         :return: The last transaction of the recordset, sorted by id.
         :rtype: recordset of `payment.transaction`
         """
-        return self.filtered(lambda t: t.state != 'draft').sorted()[:1]
+        return self and self.filtered(lambda t: t.state != 'draft').sorted()[:1] or self
```
Comment (AntoineVDV) [inline addons/payment/models/payment_transaction.py:1132]: Why did you need to change that? This seems wrong to me. The method would now not only return draft transactions, but also all transactions of the recordset (and not only the last one) when the `filtered` returns an empty recordset.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

File: addons/payment/models/payment_transaction.py (hunk 4 above)
Comment (hote-odoo) [reply addons/payment/models/payment_transaction.py:1132]: In a call to `_get_last` the recordset self might be empty (partner is then set to false) and access rules will fail.

Comment (hote-odoo) [reply addons/payment/models/payment_transaction.py:1132]: still not good, I will change it again

**Comment by pimodoo** [comment]
@AntoineVDV Hello, any idea if it will be merged before the freeze? Because the fix in saas-16.2 had not been forward-ported.

**Comment by hote-odoo** [comment]
Hello @pimodoo , I was on holiday, I'm working on it right now, should be merged today or tomorrow at worst.

File: addons/sale/models/sale_order.py [hunk 5]
```diff
@@ -1301,7 +1301,8 @@ def _notify_get_recipients_groups(self, message, model_description, msg_vals=Non
             pass
         else:
             access_opt = customer_portal_group[2].setdefault('button_access', {})
-            is_tx_pending = self.get_portal_last_transaction().state == 'pending'
+            # Only to read tx, does not produce a sudoed recordset and no information is leaked.
```
Comment (AntoineVDV) [inline addons/sale/models/sale_order.py:1304]: It's obvious when you read the code. The comment should justify using a sudo so we know why it's there, not explain that it has no consequence.
```suggestion
            #  In sudo mode to allow salesmen to read the state of the transaction.
```

File: addons/sale/models/sale_order.py [hunk 6]
```diff
@@ -1453,12 +1454,11 @@ def _has_to_be_paid(self, include_draft=False):
         :rtype: bool
         """
         self.ensure_one()
-        transaction = self.get_portal_last_transaction()
         return (
             (self.state == 'sent' or (self.state == 'draft' and include_draft))
             and not self.is_expired
             and self.require_payment
-            and transaction.state != 'done'
+            and self.sudo().get_portal_last_transaction().state != 'done'
```
Comment (AntoineVDV) [inline addons/sale/models/sale_order.py:1461]: Here too.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

**Comment by pimodoo** [comment]
> Hello @pimodoo , I was on holiday, I'm working on it right now, should be merged today or tomorrow at worst.

Thank you very much

File: addons/account_payment/views/account_portal_templates.xml [hunk 7]
```diff
@@ -115,9 +115,9 @@
             inherit_id="account.portal_invoice_success">
         <xpath expr="//a[hasclass('close')]" position="after">
             <t t-if="success == 'pay_invoice'">
-                <t t-set="payment_tx_id" t-value="invoice.get_portal_last_transaction()"/>
-                <span t-if='payment_tx_id.provider_id.sudo().done_msg' t-out="payment_tx_id.provider_id.sudo().done_msg"/>
-                <div t-if="payment_tx_id.provider_id.sudo().pending_msg and payment_tx_id.provider_code == 'custom' and invoice.ref">
+                <t t-set="tx_sudo" t-value="invoice.get_portal_last_transaction()"/>
+                <span t-if='tx_sudo.provider_id.sudo().done_msg' t-out="tx_sudo.provider_id.sudo().done_msg"/>
```
Comment (Feyensv) [inline addons/account_payment/views/account_portal_templates.xml:119]:
```suggestion
                <span t-if="tx_sudo.provider_id.sudo().done_msg" t-out="tx_sudo.provider_id.sudo().done_msg"/>
```

**Review by Feyensv — APPROVED** [review]

**Review by AntoineVDV — APPROVED** [review]

**Comment by Feyensv** [comment]
@robodoo r+

**Comment by Feyensv** [comment]
@robodoo r+

**Comment by pimodoo** [comment]
@hote-odoo @Feyensv Hello, your branch is red, could you rebase your branches, and the upgrade one (it is because of the upgrade one that you are red. Then R+ all your branches and ask @KangOl to r+ the upgrade branch.

**Comment by pimodoo** [comment]
@robodoo r+ rebase-ff
